Child logout

AryamanDubey28 opened the pull request:
Hi Guys,

This is the implementation of the child mode log out task. Log out button in child mode is clicked and then they have to enter the PIN to actually log out back to the admin home

Let me know if there's any issues!

Thanks,
Aryaman

Commit:
add text to speech functionality

Commit:
final changes

Commit:
add logout feature with PIN

Commit:
add integration tests, correct PIN and incorrect PIN

skellyhoe commented on lib/pages/admin/admin_interface.dart:
```diff
@@ -12,6 +12,12 @@ class AdminInterface extends StatelessWidget {
       theme: ThemeData(
         primarySwatch: Colors.teal,
       ),
+      // routes: {
```
delete commented out code

skellyhoe commented on lib/pages/child_board/child_board_interface.dart:
```diff
@@ -14,6 +15,13 @@ class ChildInterface extends StatelessWidget {
+      routes: {
+        // When navigating to the "homeScreen" route, build the HomeScreen widget.
+        'homeScreen': (context) => ChildBoards(),
+        'adminScreen': (context) => AdminChoiceBoards(),
```
this whole file will be deleted make sure it has nothing important

skellyhoe commented on lib/pages/child_menu/customizable_column.dart:
```diff
@@ -37,10 +58,72 @@ class CustomizableColumn extends StatelessWidget {
+        MaterialPageRoute(builder: (context) => AdminChoiceBoards()),
+      );
+
+      //Navigator.popAndPushNamed(context, 'adminScreen');
```
remove commented out code

skellyhoe commented on lib/pages/child_menu/customizable_column.dart:
```diff
@@ -37,10 +58,72 @@ class CustomizableColumn extends StatelessWidget {
+    //verifys password is correct, if so then navigates back. otherwise says incorrect
+    if (pin_controller.text.trim() == "0000") {
+      print("navigating back to admin home");
+      Navigator.push(
```
use pushAndReplace instead of push

skellyhoe commented on lib/pages/child_menu/customizable_column.dart:
```diff
@@ -37,10 +58,72 @@ class CustomizableColumn extends StatelessWidget {
+                key: Key("logoutButton"),
+                //only triggers when its pressed for some time and swiped up
+                // onLongPressUp: () {
+                //   openLogoutDialog(context);
```
commented out code

skellyhoe commented on lib/widgets/logout_alert_dialog.dart:
```diff
@@ -0,0 +1,12 @@
+import 'package:flutter/material.dart';
+import 'package:flutter/src/widgets/framework.dart';
+import 'package:flutter/src/widgets/placeholder.dart';
+
```
what does this widget do?

skellyhoe commented on lib/pages/child_menu/customizable_column.dart:
```diff
@@ -37,10 +58,72 @@ class CustomizableColumn extends StatelessWidget {
+  void submit(BuildContext context) {
+    //verifys password is correct, if so then navigates back. otherwise says incorrect
+    if (pin_controller.text.trim() == "0000") {
+      print("navigating back to admin home");
```
remove print statement

skellyhoe reviewed (changes requested):
Good work!

Some notes:

- Remove commented out code highlighted.
- Write general comments on methods and pages you created.
- Remove print statement highlighted.
- Use pushAndReplace instead of push where highlighted.
- You wrote code in child_board_interface, but that file will be deleted. Make sure if the code is needed to add it somewhere else.

For the log out button consider these changes:
- When the admin presses the activate child view for the first time it should ask the admin for a PIN and then it saves it and doesn't ask for it again.
- It's still set to a tap gesture and not a hold and swipe gesture.
- Instead of a whole keyboard, the user should be presented with a number pad. 
- Instead of a whole text box, the user should be presented with a simple dialog as explained on WhatsApp.

Commit:
Number keyboard pops up and only allows numbers to be inputted, pushReplacement instead of push included, deleted commented out code

Commit:
add functionality to stay on child mode even if app is refreshed or restarted and user has not logged out of child mode

Ali-Alkhars commented on lib/main.dart:
```diff
@@ -1,21 +1,32 @@
+      theme: ThemeData(
+        primarySwatch: Colors.teal,
+      ),
+      //home: Wrapper(),
```
Remove commented out code

Ali-Alkhars commented on lib/pages/admin/admin_side_menu.dart:
```diff
@@ -37,25 +38,36 @@ class AdminSideMenu extends StatelessWidget {
-              onTap: () => Navigator.of(context).pushReplacement(MaterialPageRoute(
-                builder: (context) => CustomizableColumn(),
-              )),
+              // onTap: () => Navigator.of(context).pushReplacement(MaterialPageRoute(
```
Remove commented out code

Ali-Alkhars commented on lib/pages/authenticate/wrapper.dart:
```diff
@@ -16,7 +19,8 @@ class Wrapper extends StatelessWidget {
             //if we are trying to sign in and snapshot contains user data, we are logged in
-            return const AdminInterface();
+            return isInChildMode ? CustomizableColumn() : AdminInterface();
+            //return const AdminInterface();
```
Remove commented out code

Ali-Alkhars commented on lib/pages/child_menu/customizable_column.dart:
```diff
@@ -37,10 +60,74 @@ class CustomizableColumn extends StatelessWidget {
+
+  Future<void> submit(BuildContext context) async {
+    //verifys password is correct, if so then navigates back. otherwise says incorrect
+    if (pin_controller.text.trim() == "0000") {
```
Why is the password hardcoded?

Ali-Alkhars commented on lib/widgets/clickable_images_grid.dart:
```diff
@@ -17,14 +18,33 @@ Expanded getMainImages(List<ClickableImage> images) {
   );
 }
 
+Future speak(String text) async {
```
Add a doc comment explaining what the method does

Ali-Alkhars reviewed (changes requested):
Hello, nice work overall!

I have some minor comments you could see below. 

**Other than that I have one comment,** I know that integration tests are cool and everything, but they're slow and kinda shi**y ngl haha. I don't really see a reason why you can't do the same tests as widget tests

Commit:
Merge remote-tracking branch 'origin/master' into child_logout

Commit:
merge main into child_logout

Commit:
pin can be updated in 'edit account' page

Commit:
pin can be updated

Commit:
new user has to make pin first before activating child mode

Commit:
add comments

Commit:
add tests and fix edit pin method to ensure pin is 4 digits

AryamanDubey28 commented:
Hi guys,
Apologies for the delays but I've incorporated all the changes so check them out when you can!

Commit:
fix updating pin bug and add better error handling messages

skellyhoe commented on lib/widgets/edit_pin_section.dart:
```diff
@@ -0,0 +1,166 @@
+    String result =
+        await authentitcationHelper.createPIN(_pinEditController.text.trim());
+    show_alert_dialog(result);
+    //Navigator.of(context).pop();
```
commented out code

skellyhoe commented on lib/widgets/edit_pin_section.dart:
```diff
@@ -0,0 +1,166 @@
+  EditPINSection({required this.authentitcationHelper});
+
+  // Displays an alert dialog with the text passed as parameter.
+  void show_alert_dialog(String text) {
```
Duplicate method. Make it a separate widget if possible.

skellyhoe commented on lib/main.dart:
```diff
@@ -27,8 +35,12 @@ class MyApp extends StatelessWidget {
+      theme: ThemeData(
+        primarySwatch: Colors.teal,
+      ),
+      //home: Wrapper(),
```
still commented out code here

skellyhoe commented on lib/pages/admin/admin_side_menu.dart:
```diff
@@ -38,7 +41,22 @@ class AdminSideMenu extends StatelessWidget {
       );
 
+  // Displays an alert dialog with the text passed as parameter.
+  void show_alert_dialog(BuildContext context, String text) {
```
Duplicate method. Make it a separate widget if possible.

skellyhoe reviewed (approved):
Nice job!

I left some comments down. For the dialog widget, if it doesn't work then don't implement it. Make sure to pull from main to get the most up to date changes and see if they conflict with your work. I'll approve it now but make sure to delete the commented out code first. Also check with the others to see if we're still doing integration tests.

Commit:
removed commented code and made showDialog into a static method

Commit:
Merge branch 'master' into child_logout

Commit:
merge main into branch

Commit:
final changes

Commit:
fix failing tests

Commit:
fix tests

Commit:
fix tests

Ali-Alkhars reviewed (changes requested):
- For some reason the themes don't work on the admin's side
- Users can enter a text pin from the edit account page and it gets accepted! (the keyboard should be only digits & it should show an error message if it was anything other than digits)
- Try clicking on the change pin text. The keyboard covers the text field.
- I wasn't asked to enter a pin. When I clicked on child app without setting a pin I was taken there and got stuck!

absirgue reviewed (changes requested):
Hi Aryaman,

Thanks for the work. I like the hold button and the new edit account page.

First (as shown by screenshots attached), firebase errors are passed in to the prompt for the pin input field in the edit account page. This should abolsutely be avoided (show a different UI when my pin is not yet configured, will be way better + actually catch those errors and never display them to the user as raw as this).

Second, the custom theme feature fails on your branch. It works on main and everywhere else, please find the reason why it does not on ur branch (probably a failed merge conflict resolution) and correct it so that the custom themes are exactly working as they do on main (just copy main from main and do some of ur changes again if needed). 

Third, the testing does not cover enough cases (as illustrated by the fact that the other points of the reviews are coming up only in manual testing). Please test:
- pins longer than 4 digits
- pins that don't only contain numbers
- you can not navigate to child app if you didn't set a pin

Fourth, there is a significant UX problem: the admin can go in the child app before having even set a pin. This does not make much sense and should be avoided (show na alert dialog if you don't have set a pin yet and u try to go to the child app asking to first set a pin).

Five, you can put letters in your pin which 1- is not what we want. 2- you can not input letters in the numeric keyboard when trying to go from child app to admin app. Please change the keyboard for pin creation to be the same numeric keyboard

I know this is not the expected review as we were all hoping to be ready to merge this task. It is now 13 days late. Please integrate all of these changes ASAP and do manual testing on your own for a significant time before asking for reviews again. Your sprint 3 task is big and this task now needs to be done in a nice way. 
<img width="603" alt="Screen Shot 2023-03-10 at 6 07 05 PM" src="https://user-images.githubusercontent.com/94190429/224393995-eba643b5-cf02-40ba-8cc5-f99405f1908b.png">
<img width="610" alt="Screen Shot 2023-03-10 at 6 07 34 PM" src="https://user-images.githubusercontent.com/94190429/224393996-eddc7ef1-014c-4f21-b6d1-110aa665acff.png">

Thanks

Commit:
fix issues:
- cant log into child mode without making a PIN
- No more showing firebase errors to user
- Only numbers can be put into PIN

Commit:
fix themes issue and add tests

dmunipi1 reviewed (approved):
Well done Aryaman

Commit:
change

Commit:
Merge branch 'master' into child_logout

Commit:
merge master into branch

Commit:
changes
